File: src/sly/speakrecognizer/audio/modules/Preemphasis.scala
package sly.speakrecognizer.audio.modules

import sly.speakrecognizer.audio.Module
import sly.speakrecognizer.audio.Data
import sly.speakrecognizer.audio.SingleData

class Preemphasis(alpha: Double = 0.97) extends Module {
	
  private var sample = 0.0
  
  override def process(input: Data) = Some(new SingleData(input.asArrayDouble(0).map{ v => sample = v - alpha * sample ; sample}))
  
}

File: java/src/sly/speakrecognizer/modules/soundanalyzer/Preemphasis.java
package sly.speakrecognizer.modules.soundanalyzer;

import sly.speakrecognizer.modules.AbstractModule;
import sly.speakrecognizer.modules.Data;
import sly.speakrecognizer.modules.Module;
import sly.speakrecognizer.modules.SoundSamplesData;

public class Preemphasis extends AbstractModule{
	double sample = 0.0;
	double alpha = 0.97;
	public Preemphasis(Module output, Parameters parameters) {
		super(output);
	}

	public void putData(Data data) {
		SoundSamplesData sdata = (SoundSamplesData) data;
		double[] tab = sdata.getData(0);

	    for (int i=0; i < tab.length; i++) {
	    	sample = tab[i] -= alpha * sample;
	    }		
	    if(output != null){
	    	output.putData(sdata);
	    }
	}
}
